### PR TITLE
sysrepocfg '--merge' option

### LINKED ISSUE #813: New feature request: '--merge' option in sysrepocfg
### Description
It would be nice to have a '--merge' option in sysrepocfg in order to merge a xml/json data file without deleting any other existing data node.
A possible syntax could be:

`sysrepocfg` --merge=merge.json --format=json module-name

If no format is provided xml format will be assumed.
For example data coming from a web browser form partially filled by the user can be merged calling sysrepocfg with --merge option as a CGI executable.

Regards
fabio

### PR DESCRIPTION
**Pull request by fabioantonini**
### Description
This pull request adds the new option '--merge' to sysrepocfg
The new option modifies existing nodes or creates new nodes according the content of the specified xml/json file.

Some examples of the command

```
sysrepocfg --merge=merge.json --format=json module-name
sysrepocfg -m merge.json module-name
```

If no format is provided xml format will be assumed.
The new option doesn't delete any other existing data node as '--import' does.

Some additional tests have been added to sysrepocfg_test

### Closure
Closes #813

### CODE REVIEW

File: src/executables/sysrepocfg.c [hunk 1]
```diff
@@ -930,6 +931,178 @@ srcfg_import_datastore(struct ly_ctx *ly_ctx, int fd_in, md_module_t *module, sr
...
+/**
+ * @brief Merge content of the specified datastore for the given module from a file
+ * referenced by the descriptor 'fd_in'
+ */
+static int
+srcfg_merge_datastore(struct ly_ctx *ly_ctx, int fd_in, md_module_t *module, srcfg_datastore_t datastore,
+                       LYD_FORMAT format, bool permanent)
+{
+    int rc = SR_ERR_INTERNAL;
+	unsigned i = 0;
```
Comment (rkrejci) [inline src/executables/sysrepocfg.c]: please fix formatting

File: src/executables/sysrepocfg.c [hunk 2]
```diff
@@ -930,6 +931,178 @@ srcfg_import_datastore(struct ly_ctx *ly_ctx, int fd_in, md_module_t *module, sr
     return rc;
 }
 
+/**
```
Comment (rkrejci) [inline src/executables/sysrepocfg.c]: `srcfg_merge_datastore()` differs from `srcfg_import_datastore()` just a little, please avoid code duplication and merge the code into a single function with some parameter/option to switch between import and merge functionality. The same applies to `srcfg_merge_operation()` and `srcfg_import_operation()`.

File: src/executables/sysrepocfg.c [hunk 3]
```diff
@@ -1440,18 +1663,21 @@ srcfg_export_xpath_operation(md_module_t *module, const char *filepath, const ch
...
-        printf("Unable to delete item. Canceling the operation.\n");
+    for (i = 0; i < xpathdel_count; i++) {
+        rc = sr_delete_item(srcfg_session, xpath[i], SR_EDIT_DEFAULT);
+        if (SR_ERR_OK != rc) {
+            srcfg_report_error(rc);
+            printf("Unable to delete item. Canceling the operation.\n");
+        }
     }
-
+
```
Comment (rkrejci) [inline src/executables/sysrepocfg.c]: remove spaces from the blank line

**Comment by fabioantonini** [comment]
Hi Radek
I cannot understand why sysrepocfg_test failed? What's wrong? 
I have run it on my target board and it was successfull.

File: src/executables/sysrepocfg.c (hunk 2 above)
Comment (fabioantonini) [inline src/executables/sysrepocfg.c]: Now both srcfg_merge_datastore() and srcfg_import_operation() have a new parameter (boot merge) to select import (merge=flase) or merge operation (merge=true)

**Comment by fabioantonini** [comment]
Analysing more in details TRAVIS CI's output I can see 

[ 69%] Building C object src/CMakeFiles/sysrepocfg.dir/executables/sysrepocfg.c.o
Linking CXX shared library libSysrepo-cpp.so
**/bin/sh: 1: ../src/sysrepoctl: not found**
make[2]: *** [swig/libSysrepo-cpp.so] Error 127
make[1]: *** [swig/CMakeFiles/Sysrepo-cpp.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
Linking C executable sysrepocfg
[ 69%] Built target sysrepocfg
make: *** [all] Error 2

it seems that sysrepoctl has been called before it has been compiled? Do you agree
thanks
fabio

**Comment by rkrejci** [comment]
yes, I've just restarted the Travis's job

**Comment by fabioantonini** [comment]
thanks
please take a look at the changes. 
The modified function 'srcfg_import_datastore' is used also in 'srcfg_edit_operation', but in that case the merge parameter is set to 'false'. At the moment the user cannot edit and merge.
Can I improve that in some way in your opinion?

File: src/executables/sysrepocfg.c [hunk 4]
```diff
@@ -855,7 +856,8 @@ srcfg_import_datastore(struct ly_ctx *ly_ctx, int fd_in, md_module_t *module, sr
         switch (diff->type[i]) {
             case LYD_DIFF_DELETED:
                 SR_LOG_DBG("<LYD_DIFF_DELETED> node: %s", first_xpath);
-                rc = srcfg_convert_lydiff_deleted(first_xpath);
+                if (!merge)
```
Comment (rkrejci) [inline src/executables/sysrepocfg.c]: Please keep even one-line if statement content in code block (`if {...}`)

File: src/executables/sysrepocfg.c [hunk 5]
```diff
@@ -1885,28 +1908,21 @@ main(int argc, char* argv[])
...
-			rc = SR_ERR_INVAL_ARG;
-			goto terminate;
-		}
-#endif
+    } else if (NULL == module_name && ((operation == SRCFG_OP_EXPORT_XPATH) || (operation == SRCFG_OP_IMPORT_XPATH))) {
+        /* module name got from xpath */
+        if (xpath) {
+            snprintf(module_name_xpath, strchr(xpath, ':') - xpath, "%s", xpath + 1);
+            module_name = module_name_xpath;
+        }
```
Comment (rkrejci) [inline src/executables/sysrepocfg.c]: please keep `}` and `else` on the same line: `} else {`

**Pull request merged**